# Mandatory Part - Q2 (a)

In [146]:
import pandas as pd
import numpy as np
import math

In [147]:
# loading preprocessed state-wise cases & deaths datasets
delaware = pd.read_csv("/Users/meet/Desktop/544_project/Mandatory_dataset/delaware.csv")
idaho = pd.read_csv("/Users/meet/Desktop/544_project/Mandatory_dataset/idaho.csv")

In [148]:
idaho.head()

tot_cases  tot_death submission_date state  new_case  pnew_case  new_death  \
0         46          2      2020-04-17    ID        46        8.0          2   
1         64          3      2020-04-20    ID        64        7.0          3   
2         30          3      2020-04-21    ID        30        9.0          3   
3         36          3      2020-04-22    ID        36        7.0          3   
4         34          0      2020-04-23    ID        34        1.0          0   

   pnew_death              created_at consent_cases consent_deaths  \
0         0.0  04/17/2020 04:22:39 PM         Agree          Agree   
1         0.0  04/20/2020 04:22:39 PM         Agree          Agree   
2         1.0  04/21/2020 04:22:39 PM         Agree          Agree   
3         1.0  04/22/2020 04:22:39 PM         Agree          Agree   
4         0.0  05/25/2020 03:38:40 PM         Agree          Agree   

   prob_cases  conf_cases  conf_death  prob_death  tot_cases_old  \
0        89.0      1566.0        29.0        14.0           1655   
1        98.0      1638.0        32.0        16.0           1736   
2       107.0      1659.0        34.0        17.0           1766   
3       114.0      1688.0        36.0        18.0           1802   
4       115.0      1721.0        36.0        18.0           1836   

   tot_death_old  
0             43  
1             48  
2             51  
3             54  
4             54

## Filtering & Preprocessing

Here, we are filtering the Covid cases & deaths dataset for the states Idaho and Delaware as required in the question.

#### Filtering Idaho state's cases and deaths dataset to retrieve February 2021 data only.

In [149]:
# filter idaho dataset for February 2021
idaho["submission_date"] = pd.to_datetime(idaho["submission_date"])
idaho_feb = idaho[(idaho["submission_date"].dt.month == 2) & (idaho["submission_date"].dt.year == 2021)]
idaho_feb.head()

tot_cases  tot_death submission_date state  new_case  pnew_case  \
251        482         10      2021-02-01    ID       482      102.0   
252        491          6      2021-02-02    ID       491      111.0   
253        507          7      2021-02-03    ID       507      114.0   
254        402         -1      2021-02-04    ID       402       94.0   
255        404         11      2021-02-05    ID       404       78.0   

     new_death  pnew_death              created_at consent_cases  \
251         10         1.0  02/02/2021 02:51:51 PM         Agree   
252          6         1.0  02/03/2021 02:55:58 PM         Agree   
253          7         0.0  02/04/2021 02:53:58 PM         Agree   
254         -1         0.0  02/05/2021 02:54:41 PM         Agree   
255         11         0.0  02/06/2021 04:05:17 PM         Agree   

    consent_deaths  prob_cases  conf_cases  conf_death  prob_death  \
251          Agree     30307.0    132858.0      1522.0       213.0   
252          Agree     30418.0    133238.0      1527.0       214.0   
253          Agree     30532.0    133631.0      1534.0       214.0   
254          Agree     30626.0    133939.0      1533.0       214.0   
255          Agree     30704.0    134265.0      1544.0       214.0   

     tot_cases_old  tot_death_old  
251         163165           1735  
252         163656           1741  
253         164163           1748  
254         164565           1747  
255         164969           1758

In [150]:
len(idaho_feb)

28

#### Filtering Idaho state's cases and deaths dataset to retrieve March 2021 data only.

In [151]:
# filter idaho dataset for March 2021
idaho["submission_date"] = pd.to_datetime(idaho["submission_date"])
idaho_march = idaho[(idaho["submission_date"].dt.month == 3) & (idaho["submission_date"].dt.year == 2021)]
idaho_march.head()

tot_cases  tot_death submission_date state  new_case  pnew_case  \
279        322          7      2021-03-01    ID       322       89.0   
280        374          4      2021-03-02    ID       374      108.0   
281        452          5      2021-03-03    ID       452      107.0   
282        299          0      2021-03-04    ID       299       95.0   
283        240          0      2021-03-05    ID       240       98.0   

     new_death  pnew_death              created_at consent_cases  \
279          7         0.0  03/02/2021 02:54:47 PM         Agree   
280          4         0.0  03/03/2021 03:20:17 PM         Agree   
281          5         1.0  03/04/2021 03:08:33 PM         Agree   
282          0         0.0  03/05/2021 04:11:33 PM         Agree   
283          0         0.0  03/06/2021 03:08:48 PM         Agree   

    consent_deaths  prob_cases  conf_cases  conf_death  prob_death  \
279          Agree     32406.0    139056.0      1641.0       226.0   
280          Agree     32514.0    139322.0      1645.0       226.0   
281          Agree     32621.0    139667.0      1649.0       227.0   
282          Agree     32716.0    139871.0      1649.0       227.0   
283          Agree     32814.0    140013.0      1649.0       227.0   

     tot_cases_old  tot_death_old  
279         171462           1867  
280         171836           1871  
281         172288           1876  
282         172587           1876  
283         172827           1876

In [152]:
len(idaho_march)

30

#### Filtering Delaware state's cases and deaths dataset to retrieve February 2021 data only.

In [153]:
# filter delaware dataset for February 2021
delaware["submission_date"] = pd.to_datetime(delaware["submission_date"])
delaware_feb = delaware[(delaware["submission_date"].dt.month == 2) & (delaware["submission_date"].dt.year == 2021)]
delaware_feb.head()

tot_cases  tot_death submission_date state  new_case  pnew_case  \
309        251          5      2021-02-04    DE       251       32.0   
310        362          0      2021-02-08    DE       362       35.0   
311        385          7      2021-02-12    DE       385       58.0   
312        344          1      2021-02-14    DE       344       38.0   
313        252          0      2021-02-15    DE       252       27.0   

     new_death  pnew_death              created_at consent_cases  \
309          5         0.0  02/06/2021 12:00:00 AM         Agree   
310          0         0.0  02/10/2021 12:00:00 AM         Agree   
311          7         0.0  02/14/2021 12:00:00 AM         Agree   
312          1         0.0  02/16/2021 12:00:00 AM         Agree   
313          0         0.0  02/17/2021 12:00:00 AM         Agree   

    consent_deaths  prob_cases  conf_cases  conf_death  prob_death  \
309          Agree      3728.0     76425.0      1022.0       113.0   
310          Agree      3837.0     78106.0      1091.0       117.0   
311          Agree      4008.0     79314.0      1148.0       121.0   
312          Agree      4069.0     80024.0      1160.0       123.0   
313          Agree      4096.0     80249.0      1160.0       123.0   

     tot_cases_old  tot_death_old  
309          80153           1135  
310          81943           1208  
311          83322           1269  
312          84093           1283  
313          84345           1283

In [154]:
len(delaware_feb)

10

#### Filtering Idaho state's cases and deaths dataset to retrieve March 2021 data only.

In [155]:
# filter delaware dataset for February 2021
delaware["submission_date"] = pd.to_datetime(delaware["submission_date"])
delaware_march = delaware[(delaware["submission_date"].dt.month == 3) & (delaware["submission_date"].dt.year == 2021)]
delaware_march.head()

tot_cases  tot_death submission_date state  new_case  pnew_case  \
319        290          0      2021-03-01    DE       290       22.0   
320        136          4      2021-03-02    DE       136       16.0   
321        229          4      2021-03-04    DE       229       29.0   
322        219          9      2021-03-05    DE       219        6.0   
323        232          9      2021-03-07    DE       232       16.0   

     new_death  pnew_death              created_at consent_cases  \
319          0         0.0  03/03/2021 12:00:00 AM         Agree   
320          4         0.0  03/04/2021 12:00:00 AM         Agree   
321          4         0.0  03/06/2021 12:00:00 AM         Agree   
322          9         0.0  03/07/2021 12:00:00 AM         Agree   
323          9         0.0  03/09/2021 12:00:00 AM         Agree   

    consent_deaths  prob_cases  conf_cases  conf_death  prob_death  \
319          Agree      4393.0     83820.0      1286.0       136.0   
320          Agree      4409.0     83940.0      1290.0       136.0   
321          Agree      4459.0     84370.0      1308.0       136.0   
322          Agree      4465.0     84583.0      1317.0       136.0   
323          Agree      4508.0     85022.0      1337.0       136.0   

     tot_cases_old  tot_death_old  
319          88213           1422  
320          88349           1426  
321          88829           1444  
322          89048           1453  
323          89530           1473

In [156]:
len(delaware_march)

24

# Wald's Test

In [161]:
import pandas as pd
import numpy as np
import collections
import csv
import math
import warnings
warnings.filterwarnings('ignore')

In [162]:
# Reading CSV
vaccines = pd.read_csv('/Users/meet/Desktop/544_project/Mandatory_dataset/COVID-19_Vaccinations_in_the_United_States_Jurisdiction 2.csv')

In [163]:
vaccines.head()

Date  MMWR_week Location  Distributed  Distributed_Janssen  \
0  05/15/2022         20       PR      7552350               215000   
1  05/15/2022         20       KS      6121515               256400   
2  05/15/2022         20       VA     19949085               785300   
3  05/15/2022         20       MT      2004895               105200   
4  05/15/2022         20      IH2      2965895               108400   

   Distributed_Moderna  Distributed_Pfizer  Distributed_Unk_Manuf  \
0              2662120             4675230                      0   
1              2354940             3510175                      0   
2              7108700            12055085                      0   
3               825200             1074495                      0   
4              1311680             1545815                      0   

   Dist_Per_100K  Distributed_Per_100k_12Plus  ...  \
0         236477                       265154  ...   
1         210122                       249685  ...   
2         233718                       273014  ...   
3         187588                       218494  ...   
4         142399                            0  ...   

   Administered_Dose1_Recip_5PlusPop_Pct  Series_Complete_5Plus  \
0                                   95.0              2656609.0   
1                                   79.8              1796846.0   
2                                   91.3              6263803.0   
3                                   69.3               608133.0   
4                                    0.0               873824.0   

   Series_Complete_5PlusPop_Pct  Administered_5Plus  Admin_Per_100k_5Plus  \
0                          86.4           7210952.0              234410.0   
1                          65.9           4651023.0              170493.0   
2                          78.0          16635530.0              207166.0   
3                          60.4           1602572.0              159045.0   
4                           0.0           2255839.0                   0.0   

   Distributed_Per_100k_5Plus  Series_Complete_Moderna_5Plus  \
0                    245508.0                       955633.0   
1                    224397.0                       639434.0   
2                    248431.0                      2028444.0   
3                    198973.0                       238123.0   
4                         0.0                       394920.0   

   Series_Complete_Pfizer_5Plus  Series_Complete_Janssen_5Plus  \
0                     1561379.0                       139478.0   
1                     1029902.0                       126716.0   
2                     3775026.0                       456465.0   
3                      314353.0                        55131.0   
4                      440195.0                        38703.0   

   Series_Complete_Unk_Manuf_5Plus  
0                            119.0  
1                            794.0  
2                           3868.0  
3                            526.0  
4                              6.0  

[5 rows x 82 columns]

## Wald's one sample test Function ##

In [164]:
# Function for one sample Wald's test
def walds(df,col):
    df["submission_date"] = pd.to_datetime(df["submission_date"])
    df_feb = df[(df["submission_date"].dt.month == 2) & (df["submission_date"].dt.year == 2021)]
    df_mar = df[(df["submission_date"].dt.month == 3) & (df["submission_date"].dt.year == 2021)]
    guess = df_feb[col].mean()

    se = np.sqrt(df_mar[col].mean()/len(df_mar))
    
    
    theta_cap = df_mar[col].mean() #for poisson's distribution MLE is mean
    
    w = (theta_cap - guess)/se
    
    print('w statistic is ',w)
    
    # alpha = 0.05
    
    z_a2 = 1.96
    
    if abs(w)>z_a2:
        print('We reject the null hypothesis as the absolute value of Wald\'s stastic is greater than the Z value(at 0.025) = 1.96')
    else:
        print('We accept the null hypothesis as the absolute value of Wald\'s stastic is lesser than the Z value(at 0.025) = 1.96')

In [165]:
#code
print("Results of Wald's one sample test for cases for DE state")
walds(delaware,'tot_cases')
print('\n')
print("Results of Wald's one sample test for cases for ID state")
walds(idaho,'tot_cases')
print('\n')
print("Results of Wald's one sample test for deaths for DE state")
walds(delaware,'tot_death')
print('\n')
print("Results of Wald's one sample test for deaths for ID state")
walds(idaho,'tot_death')
print('\n')

Results of Wald's one sample test for cases for DE state
w statistic is  -24.805591076189682
We reject the null hypothesis as the absolute value of Wald's stastic is greater than the Z value(at 0.025) = 1.96


Results of Wald's one sample test for cases for ID state
w statistic is  -0.1584623630185989
We accept the null hypothesis as the absolute value of Wald's stastic is lesser than the Z value(at 0.025) = 1.96


Results of Wald's one sample test for deaths for DE state
w statistic is  2.0948917462655263
We reject the null hypothesis as the absolute value of Wald's stastic is greater than the Z value(at 0.025) = 1.96


Results of Wald's one sample test for deaths for ID state
w statistic is  -5.89813106089253
We reject the null hypothesis as the absolute value of Wald's stastic is greater than the Z value(at 0.025) = 1.96




## Results of Wald's one sample test for mean of cases and deaths

**Null hypothesis (H0):**

Mean of cases/deaths in the month of March'21 is equal to the mean of cases/deaths in the month of February'21 

**Alternate hypothesis (H1):**

Mean of cases/deaths in the month of March'21 is NOT equal to the mean of cases/deaths in the month of February'21 

**Result:**

**Results of Wald's one sample test for cases for DE state:**

We reject the null hypothesis as the absolute value of Wald's stastic is greater than the Z value(at 0.025) = 1.96


**Results of Wald's one sample test for cases for ID state:**

We accept the null hypothesis as the absolute value of Wald's stastic is lesser than the Z value(at 0.025) = 1.96


**Results of Wald's one sample test for deaths for DE state:**

We reject the null hypothesis as the absolute value of Wald's stastic is greater than the Z value(at 0.025) = 1.96


**Results of Wald's one sample test for deaths for ID state:**

We reject the null hypothesis as the absolute value of Wald's stastic is greater than the Z value(at 0.025) = 1.96


**Is the test applicable:**

We have used MLE as estimator which is Asymptotically normal (as n goes to infinity). Since this is the required condition, the test is applicable.  

## Wald's two sample test Function ##

In [166]:
# Function for two sample Wald's test
def walds_twot(df,col):
    df["submission_date"] = pd.to_datetime(df["submission_date"])
    df_feb = df[(df["submission_date"].dt.month == 2) & (df["submission_date"].dt.year == 2021)]
    df_mar = df[(df["submission_date"].dt.month == 3) & (df["submission_date"].dt.year == 2021)]
    theta1 = df_feb[col].mean()
    theta2 = df_mar[col].mean()

    theta_cap = theta1 - theta2
    
    guess = 0 

    se_cap = np.sqrt(df_feb[col].mean() /len(df_feb) + df_mar[col].mean() /len(df_mar))
    
    w = (theta_cap - guess)/se_cap
    
    print('w statistic is ',w)
    
    # alpha = 0.05
    
    z_a2 = 1.96
    
    if abs(w)>z_a2:
        print('We reject the null hypothesis as the absolute value of Wald\'s stastic is greater than the Z value(at 0.025) = 1.96')
    else:
        print('We accept the null hypothesis as the absolute value of Wald\'s stastic is lesser than the Z value(at 0.025) = 1.96')

In [167]:
print("Results of Wald's two sample test for cases for DE state")
walds_twot(delaware,'tot_cases')
print('\n')
print("Results of Wald's two sample test for cases for ID state")
walds_twot(idaho,'tot_cases')
print('\n')
print("Results of Wald's two sample test for deaths for DE state")
walds_twot(delaware,'tot_death')
print('\n')
print("Results of Wald's two sample test for deaths for ID state")
walds_twot(idaho,'tot_death')
print('\n')

Results of Wald's two sample test for cases for DE state
w statistic is  12.089401682022904
We reject the null hypothesis as the absolute value of Wald's stastic is greater than the Z value(at 0.025) = 1.96


Results of Wald's two sample test for cases for ID state
w statistic is  0.1100535602293159
We accept the null hypothesis as the absolute value of Wald's stastic is lesser than the Z value(at 0.025) = 1.96


Results of Wald's two sample test for deaths for DE state
w statistic is  -1.2405958875929843
We accept the null hypothesis as the absolute value of Wald's stastic is lesser than the Z value(at 0.025) = 1.96


Results of Wald's two sample test for deaths for ID state
w statistic is  3.5623551365738373
We reject the null hypothesis as the absolute value of Wald's stastic is greater than the Z value(at 0.025) = 1.96




## Results of Wald's two sample test for mean of cases and deaths

**Null hypothesis (H0):**

Mean of cases/deaths in the month of March'21 is equal to the mean of cases/deaths in the month of February'21 

**Alternate hypothesis (H1):**

Mean of cases/deaths in the month of March'21 is NOT equal to the mean of cases/deaths in the month of February'21 

**Result:**

**Results of Wald's two sample test for cases for DE state:**

We reject the null hypothesis as the absolute value of Wald's stastic is greater than the Z value(at 0.025) = 1.96


**Results of Wald's two sample test for cases for ID state:**

We accept the null hypothesis as the absolute value of Wald's stastic is lesser than the Z value(at 0.025) = 1.96


**Results of Wald's two sample test for deaths for DE state:**

We accept the null hypothesis as the absolute value of Wald's stastic is lesser than the Z value(at 0.025) = 1.96


**Results of Wald's two sample test for deaths for ID state:**

We reject the null hypothesis as the absolute value of Wald's stastic is greater than the Z value(at 0.025) = 1.96


**Is the test applicable:**

We have used MLE as estimator which is Asymptotically normal (as n goes to infinity). Since this is the required condition, the test is applicable. 

# Z - Test

### Performing Z-Test on the filtered data for February and March, as asked in the question

## Z - Test on Daily Total Cases Recorded in Idaho

In [157]:
# z-test on idaho daily total cases
print("Null Hypothesis H0: Mean of daily cases for March 2021 == Mean of daily cases for February 2021")
print("Alternate Hypothesis H1: Mean of daily cases for March 2021 != Mean of daily cases for February 2021")

sample_mean_feb_idaho = np.mean(idaho_feb["tot_cases"])
sample_mean_march_idaho = np.mean(idaho_march["tot_cases"])
sample_mean_full_idaho = np.mean(idaho["tot_cases"])

denom_cases = np.sqrt((np.sum(np.square(idaho["tot_cases"] - sample_mean_full_idaho)) / len(idaho)) / len(idaho_march))

numerator_cases = abs(sample_mean_march_idaho - sample_mean_feb_idaho)

z_stat_cases = numerator_cases / denom_cases

z_threshold = 1.96
print()
if z_stat_cases > z_threshold:
    print("Z statistic = ", z_stat_cases, "is greater than Z alpha/2 = 1.96")
    print("Hence, Null Hypothesis H0 is rejected and Alternate Hypothesis H1 is accepted,")
    print("i.e. Mean of daily cases for March 2021 != Mean of daily cases for February 2021")
else:
    print("Z statistic = ", z_stat_cases, "is less than Z alpha/2 = 1.96")
    print("Hence, Null Hypothesis H0 is accepted,")
    print("i.e. Mean of daily cases for March 2021 == Mean of daily cases for February 2021")


Null Hypothesis H0: Mean of daily cases for March 2021 == Mean of daily cases for February 2021
Alternate Hypothesis H1: Mean of daily cases for March 2021 != Mean of daily cases for February 2021

Z statistic =  0.006766790033978814 is less than Z alpha/2 = 1.96
Hence, Null Hypothesis H0 is accepted,
i.e. Mean of daily cases for March 2021 == Mean of daily cases for February 2021


## Z - Test on Daily Total Deaths Recorded in Idaho

In [158]:
# z-test on idaho daily total deaths

print("Null Hypothesis H0: Mean of daily deaths for March 2021 == Mean of daily deaths for February 2021")
print("Alternate Hypothesis H1: Mean of daily deaths for March 2021 != Mean of daily deaths for February 2021")

sample_mean_feb_idaho_death = np.mean(idaho_feb["tot_death"])
sample_mean_march_idaho_death = np.mean(idaho_march["tot_death"])
sample_mean_full_idaho_death = np.mean(idaho["tot_death"])

denom_deaths = np.sqrt((np.sum(np.square(idaho["tot_death"] - sample_mean_full_idaho_death)) / len(idaho)) / len(idaho_march))

numerator_deaths = abs(sample_mean_march_idaho_death - sample_mean_feb_idaho_death)

z_stat_deaths = numerator_deaths / denom_deaths

print()
if z_stat_deaths > z_threshold:
    print("Z statistic = ", z_stat_deaths, "is greater than Z alpha/2 = 1.96")
    print("Hence, Null Hypothesis H0 is rejected and Alternate Hypothesis H1 is accepted,")
    print("i.e. Mean of daily deaths for March 2021 != Mean of daily deaths for February 2021")
else:
    print("Z statistic = ", z_stat_deaths, "is less than Z alpha/2 = 1.96")
    print("Hence, Null Hypothesis H0 is accepted,")
    print("i.e. Mean of daily deaths for March 2021 == Mean of daily deaths for February 2021")


Null Hypothesis H0: Mean of daily deaths for March 2021 == Mean of daily deaths for February 2021
Alternate Hypothesis H1: Mean of daily deaths for March 2021 != Mean of daily deaths for February 2021

Z statistic =  2.049002692157555 is greater than Z alpha/2 = 1.96
Hence, Null Hypothesis H0 is rejected and Alternate Hypothesis H1 is accepted,
i.e. Mean of daily deaths for March 2021 != Mean of daily deaths for February 2021


## Z - Test on Daily Total Cases Recorded in Delaware

In [159]:
# z-test on delaware cases feb-march

print("Null Hypothesis H0: Mean of daily cases for March 2021 == Mean of daily cases for February 2021")
print("Alternate Hypothesis H1: Mean of daily cases for March 2021 != Mean of daily cases for February 2021")

sample_mean_feb_delaware = np.mean(delaware_feb["tot_cases"])
sample_mean_march_delaware = np.mean(delaware_march["tot_cases"])
sample_mean_full_delaware = np.mean(delaware["tot_cases"])

denom_delaware = np.sqrt((np.sum(np.square(delaware["tot_cases"] - sample_mean_full_delaware)) / len(delaware)) / len(delaware_march))

numerator_delaware = abs(sample_mean_march_delaware - sample_mean_feb_delaware)

z_stat_delaware = numerator_delaware / denom_delaware

print()
if z_stat_delaware > z_threshold:
    print("Z statistic = ", z_stat_delaware, "is greater than Z alpha/2 = 1.96")
    print("Hence, Null Hypothesis H0 is rejected and Alternate Hypothesis H1 is accepted,")
    print("i.e. Mean of daily cases for March 2021 != Mean of daily cases for February 2021")
else:
    print("Z statistic = ", z_stat_delaware, "is less than Z alpha/2 = 1.96")
    print("Hence, Null Hypothesis H0 is accepted,")
    print("i.e. Mean of daily cases for March 2021 == Mean of daily cases for February 2021")

Null Hypothesis H0: Mean of daily cases for March 2021 == Mean of daily cases for February 2021
Alternate Hypothesis H1: Mean of daily cases for March 2021 != Mean of daily cases for February 2021

Z statistic =  1.9501906849818513 is less than Z alpha/2 = 1.96
Hence, Null Hypothesis H0 is accepted,
i.e. Mean of daily cases for March 2021 == Mean of daily cases for February 2021


## Z - Test on Daily Total Deaths Recorded in Delaware

In [160]:
# z-test on delaware deaths feb-march

print("Null Hypothesis H0: Mean of daily deaths for March 2021 == Mean of daily deaths for February 2021")
print("Alternate Hypothesis H1: Mean of daily deaths for March 2021 != Mean of daily deaths for February 2021")

sample_mean_feb_delaware_death = np.mean(delaware_feb["tot_death"])
sample_mean_march_delaware_death = np.mean(delaware_march["tot_death"])
sample_mean_full_delaware_death = np.mean(delaware["tot_death"])

denom_delaware_deaths = np.sqrt((np.sum(np.square(delaware["tot_death"] - sample_mean_full_delaware_death)) / len(delaware)) / len(delaware_march))

numerator_delaware_deaths = abs(sample_mean_march_delaware_death - sample_mean_feb_delaware_death)

z_stat_delaware_deaths = numerator_delaware_deaths / denom_delaware_deaths

print()
if z_stat_delaware_deaths > z_threshold:
    print("Z statistic = ", z_stat_delaware_deaths, "is greater than Z alpha/2 = 1.96")
    print("Hence, Null Hypothesis H0 is rejected and Alternate Hypothesis H1 is accepted,")
    print("i.e. Mean of daily deaths for March 2021 != Mean of daily deaths for February 2021")
else:
    print("Z statistic = ", z_stat_delaware_deaths, "is less than Z alpha/2 = 1.96")
    print("Hence, Null Hypothesis H0 is accepted,")
    print("i.e. Mean of daily deaths for March 2021 == Mean of daily deaths for February 2021")

Null Hypothesis H0: Mean of daily deaths for March 2021 == Mean of daily deaths for February 2021
Alternate Hypothesis H1: Mean of daily deaths for March 2021 != Mean of daily deaths for February 2021

Z statistic =  1.6950732971676468 is less than Z alpha/2 = 1.96
Hence, Null Hypothesis H0 is accepted,
i.e. Mean of daily deaths for March 2021 == Mean of daily deaths for February 2021


## Results for one sample Z - Test on daily cases and deaths in the states Idaho and Delaware for the months February and March 2021

### 1. Results for one sample Z - Test on daily total cases recorded in Idaho in February-March 2021
**Null Hypothesis H0:** Mean of daily cases for March 2021 == Mean of daily cases for February 2021
**Alternate Hypothesis H1:** Mean of daily cases for March 2021 != Mean of daily cases for February 2021

It can be inferred from the acquired **z-statistic (= 0.068)** that the Null Hypothesis **H0 will be accepted** as the stat is less than the threshold (z-alpha/2 = 1.96).

Hence, The mean of daily cases recorded in the month of February 2021 in Idaho is **similar** to that recorded in the month of March 2021.

### 2. Results for one sample Z - Test on daily total deaths recorded in Idaho in February-March 2021
**Null Hypothesis H0:** Mean of daily deaths for March 2021 == Mean of daily deaths for February 2021
**Alternate Hypothesis H1:** Mean of daily deaths for March 2021 != Mean of daily deaths for February 2021

It can be inferred from the acquired **z-statistic (= 2.05)** that the Null Hypothesis **H0 will NOT be accepted** as the stat is greater than the threshold (z-alpha/2 = 1.96).

Hence, The mean of daily deaths recorded in the month of February 2021 in Idaho is **NOT similar** to that recorded in the month of March 2021.

### 3. Results for one sample Z - Test on daily total cases recorded in Delaware in February-March 2021
**Null Hypothesis H0:** Mean of daily cases for March 2021 == Mean of daily cases for February 2021
**Alternate Hypothesis H1:** Mean of daily cases for March 2021 != Mean of daily cases for February 2021

It can be inferred from the acquired **z-statistic (= 1.95)** that the Null Hypothesis **H0 will be accepted** as the stat is less than (very close but still less) the threshold (z-alpha/2 = 1.96).

Hence, The mean of daily cases recorded in the month of February 2021 in Delaware is **similar** to that recorded in the month of March 2021.

### 4. Results for one sample Z - Test on daily total deaths recorded in Delaware in February-March 2021
**Null Hypothesis H0:** Mean of daily deaths for March 2021 == Mean of daily deaths for February 2021
**Alternate Hypothesis H1:** Mean of daily deaths for March 2021 != Mean of daily deaths for February 2021

It can be inferred from the acquired **z-statistic (= 1.69)** that the Null Hypothesis **H0 will be accepted** as the stat is less than (very close but still less) the threshold (z-alpha/2 = 1.96).

Hence, The mean of daily deaths recorded in the month of February 2021 in Delaware is **similar** to that recorded in the month of March 2021.

### Z - Stat Formula
**Z = (Xbar - Mu0) / (sigma / sqrt(n))**
Here, we have used the mean cases/deaths of February 2021 as the guessed mean for verifying the Z-Test on mean cases/deaths for March 2021. Additionally, for calculating the standard deviation, we are considering the entire state-wise data instead of using the month-wise seperated data.

### Is Z - Test applicable?
Here, we are calculating the true variance for the entire state's data whereas performing the Z-Test on February-March months only, where the number of rows (n) in all 4 of the cases is <=30. Since n is small (<=30) here, Z-Test will not be reliable / applicable for this scenario.

# T test

## One sample T test Function ##

In [168]:
# Function for one sample T test
def ttest(df,col):
    df["submission_date"] = pd.to_datetime(df["submission_date"])
    df_feb = df[(df["submission_date"].dt.month == 2) & (df["submission_date"].dt.year == 2021)]
    df_mar = df[(df["submission_date"].dt.month == 3) & (df["submission_date"].dt.year == 2021)]
    n = 31
    guess = df_feb[col].mean()
    
    mar_mean = df_mar[col].mean()
    root_se = np.sqrt(np.sum(np.square(df_mar[col]-mar_mean))/len(df_mar)-1)
    
    
    denominator = (root_se/math.sqrt(n))
    t_statistic = (mar_mean - guess)/denominator
    
    
    print('t statistic is ',t_statistic)
    
    # alpha = 0.05
    
    t_tab_val = 2.042
    
    if abs(t_statistic)>t_tab_val:
        print('We REJECT the null hypothesis as the absolute value of T stastic is greater than T(30,0.025) = 2.042')
    else:
        print('We ACCEPT the null hypothesis as the absolute value of T stastic is greater than T(30,0.025) = 2.042')

In [169]:
print("Results of one sample T test for cases for DE state")
ttest(delaware,'tot_cases')
print('\n')
print("Results of one sample T test for cases for ID state")
ttest(idaho,'tot_cases')
print('\n')
print("Results of one sample T test for deaths for DE state")
ttest(delaware,'tot_death')
print('\n')
print("Results of one sample T test for deaths for ID state")
ttest(idaho,'tot_death')

Results of one sample T test for cases for DE state
t statistic is  -6.26153336851101
We REJECT the null hypothesis as the absolute value of T stastic is greater than T(30,0.025) = 2.042


Results of one sample T test for cases for ID state
t statistic is  -0.019648179891557177
We ACCEPT the null hypothesis as the absolute value of T stastic is greater than T(30,0.025) = 2.042


Results of one sample T test for deaths for DE state
t statistic is  1.5914666058220657
We ACCEPT the null hypothesis as the absolute value of T stastic is greater than T(30,0.025) = 2.042


Results of one sample T test for deaths for ID state
t statistic is  -3.77112175944854
We REJECT the null hypothesis as the absolute value of T stastic is greater than T(30,0.025) = 2.042


## Results of one sample T test for mean of cases and deaths

**Null hypothesis (H0):**

Mean of cases/deaths in the month of March'21 is equal to the mean of cases/deaths in the month of February'21 

**Alternate hypothesis (H1):**

Mean of cases/deaths in the month of March'21 is NOT equal to the mean of cases/deaths in the month of February'21 

**Result:**

**Results of one sample T test for cases for DE state**

We REJECT the null hypothesis as the absolute value of T stastic is greater than T(30,0.025) = 2.042


**Results of one sample T test for cases for ID state**

We ACCEPT the null hypothesis as the absolute value of T stastic is greater than T(30,0.025) = 2.042


**Results of one sample T test for deaths for DE state**

We ACCEPT the null hypothesis as the absolute value of T stastic is greater than T(30,0.025) = 2.042


**Results of one sample T test for deaths for ID state**

We REJECT the null hypothesis as the absolute value of T stastic is greater than T(30,0.025) = 2.042


**Is the test applicable:**

T test is applicable as the number of samples is high, using CLT, the mean becomes Asymptotically normal.

## Two sample T test Function ##

In [170]:
# Function for two sample T test
def ttest_twot(df,col):
    df["submission_date"] = pd.to_datetime(df["submission_date"])
    df_feb = df[(df["submission_date"].dt.month == 2) & (df["submission_date"].dt.year == 2021)]
    df_mar = df[(df["submission_date"].dt.month == 3) & (df["submission_date"].dt.year == 2021)]
    guess = 0
    
    numerator = df_feb[col].mean()-df_mar[col].mean()
    
    corrected_se_feb = np.sum(np.square(df_feb[col]-df_feb[col].mean()))
    corrected_se_feb = corrected_se_feb/(len(df_feb)-1)
    corrected_se_mar = np.sum(np.square(df_mar[col]-df_mar[col].mean()))
    corrected_se_mar = corrected_se_mar/(len(df_mar)-1)
    
    
    denominator = np.sqrt((corrected_se_feb/len(df_feb))+(corrected_se_mar/len(df_mar)))
    t_statistic = numerator/denominator
    
    
    print('t statistic is ',t_statistic)
    
    # alpha = 0.05
    
    t_tab_val = 2.0025
    
    if abs(t_statistic)>t_tab_val:
        print('We REJECT the null hypothesis as the absolute value of T stastic is greater than T(57,0.025) = 2.0025')
    else:
        print('We ACCEPT the null hypothesis as the absolute value of T stastic is greater than T(57,0.025) = 2.0025')

In [171]:
print('Results of two sample T test for cases for DE state')
ttest_twot(delaware,'tot_cases')
print('\n')
print('Results of two sample T test for deaths for ID state')
ttest_twot(idaho,'tot_cases')
print('\n')
print('Results of two sample T test for cases for DE state')
ttest_twot(delaware,'tot_death')
print('\n')
print('Results of two sample T test for deaths for ID state')
ttest_twot(idaho,'tot_death')

Results of two sample T test for cases for DE state
t statistic is  2.4220780179024572
We REJECT the null hypothesis as the absolute value of T stastic is greater than T(57,0.025) = 2.0025


Results of two sample T test for deaths for ID state
t statistic is  0.012242143269691952
We ACCEPT the null hypothesis as the absolute value of T stastic is greater than T(57,0.025) = 2.0025


Results of two sample T test for cases for DE state
t statistic is  -0.6967059541905922
We ACCEPT the null hypothesis as the absolute value of T stastic is greater than T(57,0.025) = 2.0025


Results of two sample T test for deaths for ID state
t statistic is  1.7578410010137446
We ACCEPT the null hypothesis as the absolute value of T stastic is greater than T(57,0.025) = 2.0025


## Results of two sample T test for mean of cases and deaths

**Null hypothesis (H0):**

Mean of cases/deaths in the month of March'21 is equal to the mean of cases/deaths in the month of February'21 

**Alternate hypothesis (H1):**

Mean of cases/deaths in the month of March'21 is NOT equal to the mean of cases/deaths in the month of February'21 

**Result:**

**Results of two sample T test for cases for DE state:**

We REJECT the null hypothesis as the absolute value of T stastic is greater than T(57,0.025) = 2.0025


**Results of two sample T test for deaths for ID state:**

We ACCEPT the null hypothesis as the absolute value of T stastic is greater than T(57,0.025) = 2.0025


**Results of two sample T test for cases for DE state:**

We ACCEPT the null hypothesis as the absolute value of T stastic is greater than T(57,0.025) = 2.0025


**Results of two sample T test for deaths for ID state:**

We ACCEPT the null hypothesis as the absolute value of T stastic is greater than T(57,0.025) = 2.0025

**Is the test applicable:**

T test is applicable as the number of samples is high, using CLT, the mean becomes Asymptotically normal.